crates/xline/src/storage/kv_store.rs

```diff
@@ -811,20 +962,119 @@ impl KvStore {
-        let event = Event {
+
+        txn_db.write_op(WriteOp::PutKeyValue(new_rev.as_revision(), kv.clone()))?;
+        *sub_revision = sub_revision.overflow_add(1);
+
+        Ok(vec![Event {
             #[allow(clippy::as_conversions)] // This cast is always valid
             r#type: EventType::Put as i32,
             kv: Some(kv),
             prev_kv: None,
+        }])
+    }
+
+    /// Handle `DeleteRangeRequest`
+    fn sync_delete_range<T>(
+        &self,
+        txn_db: &T,
+        index: &Index,
+        req: &DeleteRangeRequest,
+        revision: i64,
+        sub_revision: &mut i64,
+    ) -> Result<Vec<Event>, ExecuteError>
+    where
+        T: XlineStorageOps,
+    {
+        let keys = Self::delete_keys(
+            txn_db,
+            index,
+            &req.key,
+            &req.range_end,
+            revision,
+            sub_revision,
+        )?;
+
+        Self::detach_leases(&keys, &self.lease_collection);
+
+        Ok(Self::new_deletion_events(revision, keys))
+    }
+
+    /// Handle `TxnRequest`
+    fn sync_txn<T>(
+        &self,
+        txn_db: &T,
+        index: &Index,
+        request: &TxnRequest,
+        revision: i64,
+        sub_revision: &mut i64,
+    ) -> Result<Vec<Event>, ExecuteError>
+    where
+        T: XlineStorageOps,
+    {
+        request.check_revision(self.compacted_revision(), self.revision())?;
+        let success = request
+            .compare
+            .iter()
+            .all(|compare| Self::check_compare(txn_db, index, compare));
+        tracing::info!("txn success: {success}");
+        let requests = if success {
+            request.success.iter()
+        } else {
+            request.failure.iter()
         };
-        Ok((ops, vec![event]))
+
+        let events = requests
+            .filter_map(|op| op.request.as_ref())
+            .map(|req| match *req {
+                Request::RequestRange(_) => Ok(vec![]),
+                Request::RequestTxn(ref r) => {
+                    self.sync_txn(txn_db, index, r, revision, sub_revision)
+                }
+                Request::RequestPut(ref r) => {
+                    self.sync_put(txn_db, index, r, revision, sub_revision)
+                }
+                Request::RequestDeleteRange(ref r) => {
+                    self.sync_delete_range(txn_db, index, r, revision, sub_revision)
+                }
+            })
+            .collect::<Result<Vec<_>, _>>()?
+            .into_iter()
+            .flatten()
+            .collect();
+
+        Ok(events)
     }
 
+    /// Sync `CompactionRequest` and return if kvstore is changed
+    async fn sync_compaction(&self, req: &CompactionRequest) -> Result<Vec<Event>, ExecuteError> {
+        let revision = req.revision;
+        let ops = vec![WriteOp::PutScheduledCompactRevision(revision)];
+        // TODO: Remove the physical process logic here. It's better to move into the KvServer
```

Comment:
Why is to move the physical process logic into the `KvServer` better?

Reply:
I think we should move out the wait logic and keep the sync method simple and synchronous.